### PR TITLE
Ship custom yarn

### LINKED ISSUE #3253: More helpful error message if node isn't installed
I tried installing a labextension without having node installed, and got this error message:

`FileNotFoundError: [Errno 2] No such file or directory: 'node'`

I think it'd be more helpful for users (especially those not familiar with the node/npm world, which is probably most users) if we added a check for the existence of node, and a helpful error message pointing people in the rigt direction if it doesn't exist.

### PR DESCRIPTION
**Pull request by bollwyvl**
This takes a big 🔨 to all use of `npm` (except examples, for which a developer is free to use `npm` or whatever they want) with a shiny new private `jlpm`, which of course could be something else. 

This introduces some dependencies on the order of things, i.e. you have to `python setup.py develop` or whatever before you can get to it, but there you go.

Interested to see if _any_ CI works :P

Follow up from Steve:
- This is a huge win for reproducibility, thanks @bollwyvl!
- Uses `yarn` workspaces, which make install time incredibly fast now (from 45s -> 2s for a no-op update).  See example below.  
- Also, there are no more symlinks dangling everywhere.  (e.g. `examples/notebook/node_modules/@jupyterlab` doesn't exist now.
- we ship a `yarn.js` file but mark it as `binary` so it doesn't show up in diffs (other than on github, but that might be only because the `.gitattributes` file isn't checked in yet).
- We can't yet use `--prefer-offline` or `--ignore-optional` pending upstream fixes to `yarn`
cf yarnpkg/yarn#4818 and https://github.com/yarnpkg/yarn/issues/4876
- `jupyter lab install local-directory` is now the equivalent of `jupyter lab link local-directory`
- I rewrote `commands.py` because it had become spaghetti code
- The watch commands are now streamlined so that you only build the things you need once and then the app launches when webpack first emits files
- We temporarily use `npmjs.org` as our registry pending https://github.com/lerna/lerna/issues/688
- Fixes #3253 with a better error message: "Please install nodejs using conda or the nodejs website".

This is so awesome:

```bash
$ yarn
yarn install v1.3.2
[1/4] 🔍  Resolving packages...
success Already up-to-date.
$ node scripts/ensure-buildutils.js
✨  Done in 1.72s.
```

### CODE REVIEW

**Comment by blink1073** [comment]
Holy frijoles, that is a huge line diff.

**Comment by blink1073** [comment]
Yeah Travis has slowed to a halt today.

**Comment by bollwyvl** [comment]
Well, that's what inline yarn will do I, i guess.

It wouldn't _strictly_ have to be checked into this repo... but presumably would be shipped either in the jupyterlab package, for simplicity, or independently...

**Comment by blink1073** [comment]
Maybe pack it as a tarball so we don't have meaningless line noise?

**Comment by domoritz** [comment]
Or install yarn with npm and then copy it. 

`npm install yarn` 😇

**Comment by blink1073** [comment]
True, `npm install yarn@foo`, before anything else.  We don't need the pinned version for dev installs.

**Comment by blink1073** [comment]
Later we can decide how best to ship a local version when we have the offline handling set up.

**Comment by blink1073** [comment]
So we add an entry to the `build()` function that checks for `staging/node_modules/.bin/yarn` and installs it if it isn't there.

**Comment by blink1073** [comment]
You wouldn't be able to run `yarn foo` yourself from the staging directory though.

**Comment by domoritz** [comment]
> You wouldn't be able to run yarn foo yourself from the staging directory though.

Why?

**Comment by blink1073** [comment]
Because `which yarn` would point to the system level yarn, not the one we'd like to use.

**Comment by blink1073** [comment]
So having a separate CLI command might be the best way, but I agree we should get it via `npm install yarn@foo` and not check it in to the repo.

**Comment by domoritz** [comment]
> So having a separate CLI command might be the best way, but I agree we should get it via npm install yarn@foo and not check it in to the repo.

Or if you install yarn into `node_modules`, can you not just `require` it?

**Comment by bollwyvl** [comment]
Well, it's actually running in appveyor, so i'll wait a tick before pushing anything.

Like I said, I could go either way! But I kind of like having a really locked down version and not relying on _yet another package manager_ to get our other (other) package manager.

Might also be worth investigating if a `.gitattributes` would neatly hide it from the diff:
```
jupyterlab/yarn-*.js binary
```

**Comment by blink1073** [comment]
If we specified a `~` for the yarn version, we could pick up bug fixes.

**Comment by bollwyvl** [comment]
What what? Appveyor win on the ~~first~~ second try?

I'll try .gitattributes for now...

**Comment by bollwyvl** [comment]
Gah, the `.gitattributes` didn't clean it up. Happy to investigate further, but honestly don't like relying on anything as complex as semver for the bundled package manager. My greatest fear is that an end user will try to do something, and somehow their system's npm, or yarn, or whatever, sneaks in and gives us unpredictable results. Wrapped in six layers of logs. 

Going with a `~`, even executed at build time, introduces a situation where _we just don't know_ what's going on in there... dunno...

**Comment by blink1073** [comment]
Okay, fair enough, so npm install the fixed version if it isn't there, but make sure it's there when we ship the main package?

**Comment by bollwyvl** [comment]
> Okay, fair enough, so npm install the fixed version if it isn't there, 

Or download the single-file release. Honestly, anything that reduces the chances of creating inconsistency.

> but make sure it's there when we ship the main package?

At least we agree there :)

**Comment by blink1073** [comment]
Yeah downloading sounds fair for the dev user.   If only `wget` were available everywhere...

**Comment by bollwyvl** [comment]
Travis: Coverage passed, one error in tests: has [this one](https://travis-ci.org/jupyterlab/jupyterlab/jobs/295061835#L4263) been known to flake?

```
FAILED TESTS:

  @jupyterlab/console

    ForeignHandler

      #onIOPubMessage()

        ✖ should not reject relevant iopub messages

          Firefox 56.0.0 (Linux 0.0.0)

        rejected relevant iopub message

        @webpack:///build/console/foreign.spec.js:192:0 <- build/bundle.js:85708:26

        invokeSlot@webpack:///~/@phosphor/signaling/lib/index.js:475:0 <- build/bundle.js:6419:13

        emit@webpack:///~/@phosphor/signaling/lib/index.js:433:0 <- build/bundle.js:6377:17

        Signal.prototype.emit@webpack:///~/@phosphor/signaling/lib/index.js:106:0 <- build/bundle.js:6050:9

        TestHandler.prototype.onIOPubMessage@webpack:///build/console/foreign.spec.js:56:0 <- build/bundle.js:85572:17

        invokeSlot@webpack:///~/@phosphor/signaling/lib/index.js:475:0 <- build/bundle.js:6419:13

        emit@webpack:///~/@phosphor/signaling/lib/index.js:433:0 <- build/bundle.js:6377:17

        Signal.prototype.emit@webpack:///~/@phosphor/signaling/lib/index.js:106:0 <- build/bundle.js:6050:9

        ClientSession.prototype._onIopubMessage@webpack:///packages/apputils/lib/clientsession.js:571:0 <- build/bundle.js:175794:9

        invokeSlot@webpack:///~/@phosphor/signaling/lib/index.js:475:0 <- build/bundle.js:6419:13

        emit@webpack:///~/@phosphor/signaling/lib/index.js:433:0 <- build/bundle.js:6377:17

        Signal.prototype.emit@webpack:///~/@phosphor/signaling/lib/index.js:106:0 <- build/bundle.js:6050:9

        DefaultSession.prototype.onIOPubMessage@webpack:///packages/services/lib/session/default.js:333:0 <- build/bundle.js:193724:9

        invokeSlot@webpack:///~/@phosphor/signaling/lib/index.js:475:0 <- build/bundle.js:6419:13

        emit@webpack:///~/@phosphor/signaling/lib/index.js:433:0 <- build/bundle.js:6377:17

        Signal.prototype.emit@webpack:///~/@phosphor/signaling/lib/index.js:106:0 <- build/bundle.js:6050:9

        DefaultKernel.prototype._onWSMessage@webpack:///packages/services/lib/kernel/default.js:740:0 <- build/bundle.js:191803:13

        DefaultKernel.prototype._createSocket/this._ws.onmessage@webpack:///packages/services/lib/kernel/default.js:653:45 <- build/bundle.js:191716:47
```

**Comment by bollwyvl** [comment]
re-kicked the above test (perhaps we'll see that error again!)

**Comment by blink1073** [comment]
That is an intermittent failure that only shows up in CI

**Comment by blink1073** [comment]
Code for snagging yarn:

```python
try:
    from urllib2 import urlopen
except ImportError:
    from urllib.request import urlopen

response = urlopen('https://github.com/yarnpkg/yarn/releases/download/v1.2.1/yarn-1.2.1.js')
with open('yarn-1.2.1.js', 'wb') as fid:
     fid.write(response.read(())
```

**Comment by bollwyvl** [comment]
Ended up using the tornado httpclient (because redirects)... also added `--prefer-offline`, and some (temporary) ci spelunking to figure out if the cache dirs are funky...

**Comment by blink1073** [comment]
We can use a local .yarnrc to set which yarn to use and ensure ignore-optional. 
https://yarnpkg.com/en/docs/yarnrc#toc-yarn-path
https://yarnpkg.com/en/docs/cli/config/

**Comment by blink1073** [comment]
Then you can just type yarn

**Comment by blink1073** [comment]
(if you have yarn installed globally)

**Comment by bollwyvl** [comment]
Anyhow, looking like it kinda works. 

Indeed there's some more tweaking to do...

> a local .yarnrc 

Hm, i like but don't like the .yarnrc idea, and had considered it (seeing the `.npmrc` there). It is the only place for example to get the offline mirroring in, so at some point, whatever handles that will need to be able to change that file. I don't want to end up with _another_ one-file-to-rule-them-all.

> ensure ignore-optional

Was just trying this out, can't tell if it made it any faster... it generally does a _pretty_ good job without this flag.

> set which yarn to use

If a developer felt like hacking a personal `.yarnrc` to point at whatever we call the thing we ship, that's great. But we don't want to get support requests for someone trying to use our shipped yarn (which is why it's not called yarn) to build something other random web thing... unless it's a developer writing extensions, i suppose!

> Then you can just type yarn 

Aside from those developers, I don't think most users are itching to type yarn, or do any of this toolchain stuff. The tolerance for it failing will be low before they either go back to the shipped vanilla or whatever. 

Furthermore, there is the name collision with the already well-established resource manager, yarn. That's the whole point of this, I guess...

**Comment by blink1073** [comment]
All good points.  My concern with `ignore-optional` is if someone is hand-running `jlpm` in the staging directory, they get those scary looking `node-gyp` errors.  But at least those are typically developers, not users.

**Comment by blink1073** [comment]
If you check in the yarn file with the `.gitattributes` change, does it show up in `git diff`?  It could be that the GitHub UI is the only problem.

**Comment by bollwyvl** [comment]
> they get those scary looking node-gyp errors

yeah: I think if you know staging exists, you're going to be looking at the full, raw output, and you can follow the usual google-til-you-make-it approach of debugging

> If you check in the yarn file with the .gitattributes change

Ha, just force-pushed (formerly 93e7222) with the downloaded file taken out! If it's purely a UI concern, I can take another look at checking it back in (`.gitattributes` is still changed)... as long as it's cryptographically checked, and maybe available from multiple known sources, it doesn't _really_ make anything much more secure, though slightly more reliable.

I assumed we didn't want developers to have to carry around all of the 4MB package manager "binaries" (which would not get very well packed with `.gitattributes`) anyone on the project has ever used.

Handling this at the packaging level... if we pick a yarn version, white-label it, and stick with it for a whole jupyterlab n.x.y or n.n.x release cycle, then the build/download hit won't even be that bad. As long as it's installed by the same `pip` or `conda` means, of course...

**Comment by bollwyvl** [comment]
Hooray, back to CI win. Happy to address any other stylistic concerns (command name, yarn path location, etc). Probably lots of directions it could go.

**Comment by blink1073** [comment]
Great!  I'm bugging out for the day, but will test drive locally tomorrow morning.

**Comment by blink1073** [comment]
@domoritz, we have 🚀:

image

**Comment by blink1073** [comment]
Those are opened with Vega3 and Vega-Lite 1, respectively.

**Comment by blink1073** [comment]
@bollwyvl, the only gotcha I had was needing to `pip install -e .` before anything else.  Otherwise it is smooth as butter.  Bravo!  Going to play with it more...

**Comment by blink1073** [comment]
`CONTRIBUTING.md` mentions `npm` commands.

**Comment by blink1073** [comment]
Same with some of the `docs/`.  If you have maintainer edits turned on I can handle those.

**Comment by blink1073** [comment]
I'm not sure about checking in the lock files.  Why not ship just the app one and otherwise ignore them?

**Comment by gnestor** [comment]
@blink1073 I gave this a try but of course I'm running into issues:

```bash
# checkout this branch
pip install -e .
jlpm install --ignore-optional --prefer-offline
jlpm run build
# checkout vega3 branch of jupyter-renderers
jupyter lab link packages/vega3-extension
```

I'm seeing:

```
> jlpm run build
yarn run v1.2.1
$ webpack
module.js:529
    throw err;
    ^

Error: Cannot find module 'fs.realpath'
    at Function.Module._resolveFilename (module.js:527:15)
    at Function.Module._load (module.js:476:23)
    at Module.require (module.js:568:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/Users/grant/anaconda/share/jupyter/lab/staging/node_modules/glob/glob.js:44:10)
    at Module._compile (module.js:624:30)
    at Object.Module._extensions..js (module.js:635:10)
    at Module.load (module.js:545:32)
    at tryModuleLoad (module.js:508:12)
    at Function.Module._load (module.js:500:3)
error Command failed with exit code 1.
```

**Comment by blink1073** [comment]
Pushed b5e09c6 with docs updates and fixing the `clean:slate` command.

**Comment by blink1073** [comment]
I took a stab at yarn workspaces following this [blog post](https://yarnpkg.com/blog/2017/08/02/introducing-workspaces/), but lerna hung at the bootstrapping phase.  Sigh.

**Comment by blink1073** [comment]
Yikes, this isn't good (appveyor):

```
jlpm cache list
yarn cache v1.2.1
Name Version Registry Resolved
Done in 0.05s.
del /f yarn.lock
jlpm install --ignore-optional --prefer-offline --network-concurrency 1
yarn install v1.2.1
info No lockfile found.
[1/4] Resolving packages...
error An unexpected error occurred: "ENOENT: no such file or directory, scandir 'C:\\Users\\appveyor\\AppData\\Local\\Yarn\\cache\\v1\\npm-@types'".
info If you think this is a bug, please open a bug report with the information provided in "C:\\projects\\jupyterlab\\yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
Command exited with code 1
```

**Comment by blink1073** [comment]
I tried cleaning the cache, removing the lockfile and limiting the network concurrency based on browsing some related yarn issues.

**Comment by blink1073** [comment]
Okay, removing `--prefer-offline` did the trick.

**Comment by blink1073** [comment]
:poop:, I was able to reproduce the `scandir` error locally on Windows 10.

**Comment by blink1073** [comment]
Okay, the problem was that I had accidentally added an entry to `dependencies` when updating the docs.  I file a bug: https://github.com/yarnpkg/yarn/issues/4818.

**Comment by blink1073** [comment]
@afshin and I talked about this and had these thoughts:

We should not be adding a new command name.   We should use `yarn-path` to point a local version of yarn that is downloaded when running `pip`.   Developers need a global yarn, or an alias that points to our yarn if that is a problem.  I verified that running `yarn` in one of the subfolders honors the `.yarnrc` file.

The `jupyter lab` commands can explicitly use `node yarn.js`, so users don't need a `yarn` command.  If they wanted to run explicit commands in the app dir, they can either install `yarn` or run `node yarn.js` manually.  

If we have a separate package that needs to write the `.yarnrc` file, it would have to include the `yarn-path`, so we need to keep the name of the file stable (not include the version).

**Comment by blink1073** [comment]
We also agreed that including the `yarn.lock` files since they [should be committed on all projects](https://yarnpkg.com/blog/2016/11/24/lockfiles-for-all/).

**Comment by blink1073** [comment]
I've come back around to the idea of offering jlpm but also allowing yarn to work.  That simplifies the use cases where you don't want a global yarn.

**Comment by bollwyvl** [comment]
Hi, folks: I was traveling and then sick! What do we need to push this forward? Happy to let this particular PR drop by the wayside if there's another, more appropriate way to get it in.

Checking in the lockfiles: 👍. It's pretty hard to completely ignore their behavior, and they are the only way they forward to a real offline build mechanism. They incur a one-time hit in terms of file size, but that's a _really valuable file_, as it will protect many users from mysterious 5-level-deep `^` and `~` semantics and other things that have no place keeping people from getting useful work done. Further, if a problem _does_ occur, having the `yarn.lock`, and being able to diff it from a known-good is tremendous.

I don't recall exactly the guarantee that yarn makes about its lock files, but setting a `JUPYTERLAB_YARN` or something would also have to be followed by a version check.

As for command naming: of course it doesn't really matter what it's called... I would have initially called it `jupyter labextension yarn`, but doing a full traitlets wrapping of the command seemed silly. But the goal is to be using as homogeneous and predictable build chain on the user's box as possible, and test the hell out of that, including developer and CI tooling. Encouraging `sudo npm install -g yarn` is exactly the pattern we would want to avoid.

Re: `--prefer-offline`, sorry that was a bit cavalier. In my original implementation, which went the full offline-mirror route, it actually tried, in series, `--offline`, `--prefer-offline` and finally without modification.

**Comment by domoritz** [comment]
fwiw, having the yarn.lock file in the repository has saved us many hours in Vega-Lite. I highly recommend it!

**Comment by blink1073** [comment]
Ran into [this bug](https://github.com/yarnpkg/yarn/issues/4876) so I turned off `ignore-optional`, and upgraded to `1.3.2-legacy` to avoid [this bug](https://github.com/yarnpkg/yarn/issues/4695).  I figured we could support node 4.0 for now as opposed to forcing node 6+.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -26,9 +26,9 @@ All source code is written in [TypeScript](http://www.typescriptlang.org/Handboo
 ### Installing Node.js and npm
 
 Building the JupyterLab from its GitHub source code requires Node.js version
-6+ and Node's package manager, ``npm``.
+6+.
```
Comment (bollwyvl) [inline CONTRIBUTING.md]: Should be`4+`?

**Comment by domoritz** [comment]
:tada:

**Comment by gnestor** [comment]
Thank you @blink1073 and @bollwyvl! 👏👏👏

**Pull request merged**